Optimize date period predicates in SQL queries (index usage)

nikita2206 opened the pull request:
##### Problem:
When using DateTime fields in filters in Metabase, if we are using any kinds of date units that are not `day` for `Date` or `seconds` for `DateTime`, Metabase will produce an SQL query with a WHERE clause that looks like `DATE(field_name) BETWEEN DATE(DATE_ADD(NOW(), INTERVAL -30 DAY)) AND DATE(NOW())` (this is if you're using `day` with `DateTime` column). In case with MySQL and possibly other RDBMSs this makes it impossible for the engine to use the index on `field_name` for filtering because `field_name` has to be mapped through another function so a full scan is required here.

##### Solution:
Inclusive `BETWEEN` can be emulated as `min_value <= field < (max_value + 1 unit)`.

PS Sorry for opening a not merge-ready PR, unfortunately some tests seem to be platform-dependent so I'd like to utilise your CI to have a look at what's actually broken with my changes.  

PS2 I might still need to do that same optimisation for reports with custom TZ, so far I haven't tested this use case but obviously this should affect index usage as well.

resolves #9139
resolves #4043

sbelak commented on src/metabase/query_processor/middleware/desugar.clj:
```diff
@@ -27,25 +27,28 @@
 
     [:time-interval field (n :guard #{-1}) unit (_ :guard :include-current)]
-    [:between [:datetime-field field unit] [:relative-datetime n unit] [:relative-datetime 0 unit]]
+    [:between-rhs-ex [:datetime-field field :default] [:relative-datetime n unit] [:relative-datetime 1 unit]]
```
Why `:default` bucketing?

nikita2206 replied:
Using anything else can introduce additional function calls on the `datetime-field` in the resulting SQL which in turn doesn't allow us to utilise index on that field. The idea was that `:relative-datetime` would always return full datetime as well but it would cut out the non-relevant parts (for example `DATE_FORMAT(DATE_ADD(NOW(), INTERVAL -2 MONTH), '%Y-%m-01 00:00:00')` which results in `2019-03-01 00:00:00` for `[:relative-datetime -2 :month]`). This is currently not the case though because the default `:relative-datetime`'s behaviour is to _truncate_ the date without zero-padding it. (this is the bug I'm talking about in another comment) So I opted to introducing another `:relative-datetime-padded` clause :) And now I really don't like how much new stuff this introduces. If you have some ideas please share them with me, I will probably have time to implement them. My main objective here is to reduce, but preferably eliminate, any use of SQL functions around date time fields.

sbelak commented:
Thanks for the PR & welcome to the MB family! 

Why did you decide on using the name `:between-rhs-ex` rather than retaining `:between` as we have now?

If you want to run CI against all the databases, put `[ci all]` in your commit message.

p.s. mind signing our CLA

p.p.s. the fe-tests-e2e test failure is intermittent and was fixed in recent master. You might want to merge master in.

nikita2206 commented:
Hey @sbelak , thanks for having a look.

Firstly we've found a couple of bugs with this, as it happens fixing those bugs (I haven't pushed those fixes) introduced new ones and after fixing new ones I found myself with quite a bit of newly introduced code some of which doesn't look like something that I would be willing to merge into metabase. So I need to figure out either a nicer way of achieving what I need here or to just nicely fix/dedupe what I currently have.

Do you suggest adding an `options` parameter to the `:between` clause? As it stands `:between` works on `[a; b]` kind of ranges while for my stuff to work here we need `[a; b)` kind of range so that's why I introduced another `between` clause with these semantics.

sbelak commented:
There's some overlap between this and #9796 (see Cam's comments for how it will look when finished -- overloading +/- is a better approach than what I initially did) in the sense that you'll get better support for interval arithmetics out of the MBQL box so you might want to wait for that (I'll finish it today most likely).

The bigger issue is, that we almost certainly don't want two very similar functions with mostly same behaviour but wildly different performance characteristics, so let's try and figure out what the least breaking change in terms of :between behaviour would be and then we can decide if it's worth it or not. It would be helpful if you can list the cases where results would be different (I'm coming up with a blank).

nikita2206 commented:
So just to clarify, I'll be waiting for #9796 before continuing with this. For now I've added some dirty workarounds just to make optimizations work for our company's deployment but I'll be happy to make these optimizations for upstream later.

nikita2206 commented:
@sbelak Do you happen to know if Metabase does any changes to the datetime values when they are used in `:>` or `:<` operators with Google Analytics? It seems like Metabase it's not consistent over different drivers in Metabase how some operators work: for example `date_field < '2019-01-01'` in SQL drivers results in a similar SQL query but with GA I think it will result into `end-date=2019-01-01` which is equivalent to `date_field <= '2019-01-01'` in SQL.

flamber commented:
@nikita2206 I think the GA might be the reason for 9475 and 9476 (not linking in case I'm wrong)

sbelak commented:
Yes, it uses start/end-date https://github.com/metabase/metabase/blob/master/modules/drivers/googleanalytics/src/metabase/driver/googleanalytics/query_processor.clj#L176

nikita2206 commented:
@sbelak thanks, just wanted to make sure that it’s a bug before fixing it. (`<` and `>` on datetimes with GA behave as if they are actually `<=` and `>=`)

@flamber hm, those seem to be related to time zones. It’s certainly not very obvious what kind of tz transitions are happening between the web client, backend and a DB, and how to correctly configure it though. I guess I’ll have to set up GA to test if the off by one error is there for gt/lt operators.

nikita2206 commented:
@sbelak btw I’ve found a way around some limitations (namely, GA was the only driver not supporting `date > ... AND date < ...` kinds of expressions which I fixed) and was able to get rid of `between-rhs-ex` operator. However we still do need a way to get a relative date that is padded rather than truncated, for example: `2019-04-01 00:00:00` for `[:relative-datetime -1 :month]` instead of `2019-04`. Right now I went with the new clause `:relative-datetime-padded` but I can see how it might be undesired to introduce a new clause for this. Do you have any suggestions for that? I’m not sure if it’s entirely a good idea to add an option to the `relative-datetime` (ie change its arity) as that might have us end up with too much clutter in that clause. Maybe adding a hashmap parameter to the `relative-datetime` that would store arbitrary number of auxiliary options could work here?

nikita2206 commented:
I've opened a new PR #9904 with some fixes, instead of letting them pile up here. This PR will need that other one in order to work with GA now.

nikita2206 committed:
Optimize the way Metabase uses date fields in predicates [ci all]

    This should fix the way Metabase is using date fields in predicates where it tries to
    trim them off some parts (for example removing seconds and minutes) which in turn
    makes it impossible to utilize indices in WHERE clauses. It is possible to work with
    full DateTime while still getting correct results via partially exclusive BETWEEN
    i.e. `date_field >= LHS_DATE && date_field < (RHS_DATE + 1)`

nikita2206 commented:
This is now ready for review, can you guys have a look please when you have some time?

Things I did:
 - removed `:between-rhs-ex`, using `[:and [:>= field datetime] [:< field datetime]]` expression instead
 - added `:relative-datetime-padded` clause, about the only place where we need a different treatment for `:relative-datetime` is when unit is `:year`. It is then cast to integer that represents the year and comparison is done using that integer. In order to support conversionless comparison for date fields we need a datetime to be zero-padded: `2019-01-01 00:00:00` instead of just `2019`. It might be a better idea to change behaviour of `relative-datetime` instead of adding a new clause but then I'd need to change how Metabase parses the year when doing aggregations by year which result in `SELECT YEAR(date_field)` since now it will have to be able to parse the year out of a full date string. If you have some ideas I'll be happy to hear them

All the tests that are failing right now are failing due to them being not initialized I think (they all complain that a special env var needs to be provided otherwise they just don't run and fail). That is except for SparkSQL test - I wasn't able to understand why that one fails, I believe the query that is generated is correct:
```
SELECT ... WHERE (`t1`.`date` >= CAST(from_unixtime(unix_timestamp(date_format(CAST(from_unixtime(unix_timestamp('2015-04-01T00:00:00.000Z', 'yyyy-MM-dd\'T\'HH:mm:ss.SSS\'Z\'')) AS timestamp), 'yyyy-MM-dd'), 'yyyy-MM-dd')) AS timestamp) AND `t1`.`date` < CAST(from_unixtime(unix_timestamp(date_format(CAST(from_unixtime(unix_timestamp('2015-05-02T00:00:00.000Z', 'yyyy-MM-dd\'T\'HH:mm:ss.SSS\'Z\'')) AS timestamp), 'yyyy-MM-dd'), 'yyyy-MM-dd')) AS timestamp))
```
for this MBQL filter:
```
:filter [:between [:datetime-field $date :day] "2015-04-01" "2015-05-01"]
```
It might be that SparkSQL treats timezones in some special way?

nikita2206 commented:
Okay, SparkSQL doesn't do automatic type coercion on comparison so comparing `date('2019-01-10') < timestamp('2019-01-10')` gives `true` (i.e. when two are equal then timestamp will be considered larger than date)

nikita2206 commented:
Can you guys give this a look please, when you have time @camsaul @sbelak

realrnjs commented:
@nikita2206 hi. do not support when using native sql

nikita2206 commented:
@realrnjs not sure what you mean, can you elaborate? This is supposed to be MBQL-only optimization I think

nikita2206 committed:
attempt to make the fork of metabase work with the fork of mbql

nikita2206 committed:
Replace relative-datetime-padded with relative-datetime with options

nikita2206 commented:
Blocked by https://github.com/metabase/mbql/pull/19

camsaul commented:
Hi, thanks for this PR.

We ended up implementing this same fix in #10980
